### PR TITLE
Prevent freeing a possible invalid pointer from journald

### LINKED ISSUE #30433: dockerd died unexpectedly crashreport
**Description**

dockerd died unexpectedly on rhel 7.3

**Steps to reproduce the issue:**
1. launch docker
2. ???
3. dockerd dies

**Additional information you deem important (e.g. issue happens only occasionally):**

[log.txt](https://github.com/docker/docker/files/728735/log.txt)
**Output of `docker version`:**

```
# docker version
Client:
 Version:      1.13.0
 API version:  1.25
 Go version:   go1.7.3
 Git commit:   49bf474
 Built:        Tue Jan 17 09:55:28 2017
 OS/Arch:      linux/amd64

Server:
 Version:      1.13.0
 API version:  1.25 (minimum version 1.12)
 Go version:   go1.7.3
 Git commit:   49bf474
 Built:        Tue Jan 17 09:55:28 2017
 OS/Arch:      linux/amd64
 Experimental: false

```

**Output of `docker info`:**

```
# docker info 
Containers: 14
 Running: 14
 Paused: 0
 Stopped: 0
Images: 24
Server Version: 1.13.0
Storage Driver: overlay
 Backing Filesystem: xfs
 Supports d_type: false
Logging Driver: journald
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 03e5862ec0d8d3b3f750e19fca3ee367e13c090e
runc version: 2f7393a47307a16f8cee44a37b262e8b81021e3e
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 3.10.0-514.el7.x86_64
Operating System: Red Hat Enterprise Linux Server 7.3 (Maipo)
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 23.39 GiB
Name: scedr07owltow01
ID: JFOZ:QQNR:ZUX4:2ADX:D5GX:R5Z7:4TB2:6THT:4ZOK:DZNB:PHAV:Z3RI
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Https Proxy: 10.36.131.11:8080
Username: freeseacher
Registry: https://index.docker.io/v1/
WARNING: bridge-nf-call-iptables is disabled
WARNING: bridge-nf-call-ip6tables is disabled
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: true

```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

### PR DESCRIPTION
**Pull request by mlaventure**
Signed-off-by: Kenfe-Mickael Laventure <mickael.laventure@gmail.com>

-- 

May fix https://github.com/docker/docker/issues/30433

After looking at the code, at the moment the only explanation I can muster is that somehow `sd_journal_get_cursor()` failed and left the `cursor` variable with an invalid value.

@boutros @amatsus any chance you'd be able to test this patch?

### CODE REVIEW

**Comment by tonistiigi** [comment]
SGTM

**Comment by thaJeztah** [comment]
ping @nalind PTAL

**Comment by boutros** [comment]
> @boutros @amatsus any chance you'd be able to test this patch?

Unforunatley I don't know a way to trigger the crash - it just happend at irregular intervals, days to sometimes hours. It was happened in to differet, but equally specced production servers, so we downgraded to 1.12.6 for now.

Thanks for giving this attention @mlaventure  :)

**Comment by amatsus** [comment]
I tested this patch with docker 1.13.1. I get same results.
[docker-crash-report.txt](https://github.com/docker/docker/files/793785/docker-crash-report.txt)

I also downgraded to 1.12.6 for now.

**Comment by nalind** [comment]
@thaJeztah Looks like a good correctness fix to me.

**Comment by runcom** [comment]
apparently the issue in #30433 is still happening though https://github.com/docker/docker/pull/31231#issuecomment-281690113 :(

**Comment by mlaventure** [comment]
@amatsus thanks for testing!

@tonistiigi re-reading the crash, something is weird. The stacktrace implies that the free from `drainJournal` is called when the go routine exits, which would mean *after* `sd_journal_close()` has been called.

I'll update the patch to only close the journal in `readLogs` see if that helps.

**Comment by mlaventure** [comment]
@amatsus do you mind testing that updated version? I still haven't found a way to reproduce this.

**Comment by amatsus** [comment]
@mlaventure My bad, I used the f5007f2 patch. This is old?  I use the 81630df patch, so I get a good result. Many thanks!
```
From f5007f2e8de6dc55fada9298ccd2bd34854935bb Mon Sep 17 00:00:00 2001
From: Kenfe-Mickael Laventure <mickael.laventure@gmail.com>
Date: Tue, 21 Feb 2017 11:07:57 -0800
Subject: [PATCH] Prevent freeing a possible invalid pointer from journald

Signed-off-by: Kenfe-Mickael Laventure <mickael.laventure@gmail.com>
---
 daemon/logger/journald/read.go | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/daemon/logger/journald/read.go b/daemon/logger/journald/read.go
index 541fab70..afb7b41 100644
--- a/daemon/logger/journald/read.go
+++ b/daemon/logger/journald/read.go
@@ -237,7 +237,10 @@ drain:

        // free(NULL) is safe
        C.free(unsafe.Pointer(oldCursor))
-       C.sd_journal_get_cursor(j, &cursor)
+       if C.sd_journal_get_cursor(j, &cursor) != 0 {
+               // ensure that we won't be freeing an address that's invalid
+               cursor = nil
+       }
        return cursor
 }
```

**Comment by mlaventure** [comment]
@amatsus Not at all your fault, I updated the patch after your last comment :).

Thanks for testing it, we'll wait a bit to see if it comes back for you since you seems to be easily affected

**Comment by nalind** [comment]
(@mlaventure's re-reading of the crash also made me take a closer look at the sequence of events in `followJournal`, which prompted #31263.)

File: daemon/logger/journald/read.go [hunk 1]
```diff
@@ -237,7 +237,10 @@ drain:
 
 	// free(NULL) is safe
 	C.free(unsafe.Pointer(oldCursor))
-	C.sd_journal_get_cursor(j, &cursor)
+	if C.sd_journal_get_cursor(j, &cursor) != 0 {
+		// ensure that we won't be freeing an address that's invalid
```
Comment (tiborvass) [inline daemon/logger/journald/read.go]: Must be slow, but i dont understand this. How does this prevent freeing an invalid address?

**Comment by tiborvass** [comment]
LGTM

File: daemon/logger/journald/read.go (hunk 1 above)
Comment (nalind) [inline daemon/logger/journald/read.go]: @tiborvass the man page for `sd_journal_get_cursor()` doesn't mention what happens to the cursor pointer when the function returns a failure code.  The implementation ends up calling `asprintf()`, which explicitly says the pointer's content is unspecified, so after a failure code we have to take a step to make sure that the pointer's not still uninitialized, and that it hasn't been set to some garbage value.

**Pull request merged**